### PR TITLE
made random forrest classifier from the ground up. does not use sklearn

### PR DESCRIPTION
**Pull request by Rajkanwars15**
### Describe your change:

made random forrest classifier from the ground up. does not use sklearn

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [ ] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".
Solves part of an issue #6216

### CODE REVIEW

File: machine_learning/random_forest_classifier.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    Examples:
+        >>> np.random.seed(42)
+        >>> X = np.random.rand(100, 2)
+        >>> y = (X[:, 0] + X[:, 1] > 1).astype(int)
+        >>> tree = DecisionTree(max_depth=3)
+        >>> tree.fit(X, y)
+        >>> predictions = tree.predict(np.array([[0.7, 0.3], [0.2, 0.8]]))
+    """
+
+    def __init__(self, max_depth=None):
```
[the remark this answers is not in the document]
Comment (Rajkanwars15) [reply machine_learning/random_forest_classifier.py:23]: def __init__(self, max_depth: Optional[int] = None) -> None:

File: machine_learning/random_forest_classifier.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        max_depth (Optional[int]): Maximum depth of the tree. If None, the tree grows until pure nodes or min_samples_split is reached.
+
+    Attributes:
+        tree (tuple): The decision tree structure.
+    """
+
+    def __init__(self, max_depth: Optional[int] = None) -> None:
+        self.max_depth = max_depth
+
+    def fit(self, X: List[np.ndarray], y: List[int]) -> None:
```
[the remark this answers is not in the document]
Comment (Rajkanwars15) [reply machine_learning/random_forest_classifier.py:19]: `def fit(self, features: List[np.ndarray], labels: List[int]) -> None:
    """
    Fit the decision tree to the training data.

    Parameters:
        features (List[numpy.ndarray]): The input features.
        labels (List[int]): The target labels.

    Returns:
        None

    Examples:
        >>> np.random.seed(42)
        >>> X = np.random.rand(100, 2)
        >>> y = (X[:, 0] + X[:, 1] > 1).astype(int)
        >>> tree = DecisionTree(max_depth=3)
        >>> tree.fit(X, y)
    """
    self.tree = self._build_tree(features, labels, depth=0)
`

File: machine_learning/random_forest_classifier.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        Parameters:
+            X (List[numpy.ndarray]): The input features.
+            y (List[int]): The target labels.
+
+        Returns:
+            None
+        """
+        self.tree = self._build_tree(X, y, depth=0)
+
+    def _build_tree(self, X: List[np.ndarray], y: List[int], depth: int) -> tuple:
```
[the remark this answers is not in the document]
Comment (Rajkanwars15) [reply machine_learning/random_forest_classifier.py:32]: `def _build_tree(self, features: List[np.ndarray], labels: List[int], depth: int) -> tuple:
    """
    Recursively build the decision tree.

    Parameters:
        features (List[numpy.ndarray]): The input features.
        labels (List[int]): The target labels.
        depth (int): The current depth of the tree.

    Returns:
        tuple: The decision tree structure.

    Examples:
        >>> np.random.seed(42)
        >>> X = np.random.rand(100, 2)
        >>> y = (X[:, 0] + X[:, 1] > 1).astype(int)
        >>> tree = DecisionTree(max_depth=3)
        >>> tree._build_tree(X, y, depth=0)
    """
    if depth == self.max_depth or len(np.unique(labels)) == 1:
        return (np.bincount(labels).argmax(),)
    # Rest of the code for _build_tree function...
`

File: machine_learning/random_forest_classifier.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        )
+        right_split = self._build_tree(
+            [np.array(X)[np.array(X)[:, best_split_feature] > best_split_value]],
+            [np.array(y)[np.array(X)[:, best_split_feature] > best_split_value]],
+            depth + 1,
+        )
+
+        return (best_split_feature, best_split_value, left_split, right_split)
+
+    def _calculate_gini(self, y: List[int]) -> float:
```
[the remark this answers is not in the document]
Comment (Rajkanwars15) [reply machine_learning/random_forest_classifier.py:92]: `def _calculate_gini(self, labels: List[int]) -> float:
    """
    Calculate the Gini impurity for a given set of labels.

    Parameters:
        labels (List[int]): A list of labels.

    Returns:
        float: The Gini impurity.

    Examples:
        >>> labels = [0, 0, 1, 1, 1]
        >>> tree = DecisionTree(max_depth=3)
        >>> tree._calculate_gini(labels)
    """
    if len(labels) == 0:
        return 0
    p_i = np.bincount(labels) / len(labels)
    return 1 - np.sum(p_i**2)
`

File: machine_learning/random_forest_classifier.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        Returns:
+            float: The Gini impurity.
+        """
+        if len(y) == 0:
+            return 0
+        p_i = np.bincount(y) / len(y)
+        return 1 - np.sum(p_i**2)
+
+    def predict(self, X: List[np.ndarray]) -> List[int]:
```
[the remark this answers is not in the document]
Comment (Rajkanwars15) [reply machine_learning/random_forest_classifier.py:107]: `def predict(self, features: List[np.ndarray]) -> List[int]:
    """
    Make predictions for input features.

    Parameters:
        features (List[numpy.ndarray]): The input features.

    Returns:
        List[int]: Predicted labels.

    Examples:
        >>> np.random.seed(42)
        >>> X = np.random.rand(100, 2)
        >>> y = (X[:, 0] + X[:, 1] > 1).astype(int)
        >>> tree = DecisionTree(max_depth=3)
        >>> tree.fit(X, y)
        >>> predictions = tree.predict([np.array([0.7, 0.3]), np.array([0.2, 0.8])])
    """
    return [self._predict_tree(x, self.tree) for x in features]
`

File: machine_learning/random_forest_classifier.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+        Parameters:
+            X (List[numpy.ndarray]): The input features.
+
+        Returns:
+            List[int]: Predicted labels.
+        """
+        return [self._predict_tree(x, self.tree) for x in X]
+
+    def _predict_tree(self, x: np.ndarray, tree: tuple) -> int:
```
[the remark this answers is not in the document]
Comment (Rajkanwars15) [reply machine_learning/random_forest_classifier.py:119]: `def _predict_tree(self, data_point: np.ndarray, tree: tuple) -> int:
    """
    Recursively traverse the decision tree to make predictions.

    Parameters:
        data_point (numpy.ndarray): Input features for a single data point.
        tree (tuple): The decision tree structure.

    Returns:
        int: Predicted label.

    Examples:
        >>> np.random.seed(42)
        >>> X = np.random.rand(100, 2)
        >>> y = (X[:, 0] + X[:, 1] > 1).astype(int)
        >>> tree = DecisionTree(max_depth=3)
        >>> tree.fit(X, y)
        >>> data_point = np.array([0.7, 0.3])
        >>> prediction = tree._predict_tree(data_point, tree.tree)
    """
    if len(tree) == 1:
        return tree[0]
    feature, value, left, right = tree
    if data_point[feature] <= value:
        return self._predict_tree(data_point, left)
    else:
        return self._predict_tree(data_point, right)
`
